neo4j/_async/work/workspace.py

```diff
@@ -77,6 +82,76 @@ def _set_cached_database(self, database):
         self._cached_database = True
         self._config.database = database
 
+    def _initialize_bookmarks(self, bookmarks):
+        if isinstance(bookmarks, Bookmarks):
+            prepared_bookmarks = tuple(bookmarks.raw_values)
+        elif hasattr(bookmarks, "__iter__"):
+            deprecation_warn(
+                "Passing an iterable as `bookmarks` to `Session` is "
+                "deprecated. Please use a `Bookmarks` instance.",
+                stack_level=5
+            )
+            prepared_bookmarks = tuple(bookmarks)
+        elif not bookmarks:
+            prepared_bookmarks = ()
+        else:
+            raise TypeError("Bookmarks must be an instance of Bookmarks or an "
+                            "iterable of raw bookmarks (deprecated).")
+        self._initial_bookmarks = self._bookmarks = prepared_bookmarks
+
+    async def _get_bookmarks(self, database):
+        if self._bookmark_manager is not None:
+            # For 4.3- support: the server will not send the resolved home
+            # database back. To avoid confusion between `None` as in "all
+            # database" and `None` as in "home database" we re-write the
+            # home database to `""`, which otherwise is an invalid database
+            # name.
+            if database is None:
+                database = ""
+            self._bookmarks = tuple({
+                *await AsyncUtil.callback(
+                    self._bookmark_manager.get_bookmarks, database
+                ),
+                *self._initial_bookmarks
+            })
+        return self._bookmarks
+
+    async def _get_all_bookmarks(self):
+        if self._bookmark_manager is not None:
+            self._bookmarks = tuple({
+                *await AsyncUtil.callback(
+                    self._bookmark_manager.get_all_bookmarks,
+                ),
+                *self._initial_bookmarks
+            })
+        return self._bookmarks
```

Comment:
Changing the value of `self._bookmarks` breaks the `session.last_bookmarks` contract. For instance, if you run a read query and then try to get the `session.last_bookmarks`, you will get all the bookmarks in the `BookmarkManager` instead of the bookmarks from the session.

Reply:
:scream: